### ISSUE TITLE
[Feature Request] New `BeartypeConf(is_random=False)` option forcing @beartype to type-check deterministically 😯 

### ISSUE THREAD

**Issue by leycec** [beartype/beartype#385]
### Discussed in https://github.com/beartype/beartype/discussions/383

<div type='discussions-op-text'>

<sup>Originally posted by **avolchek** May 23, 2024</sup>
Hi!

The documentation states that for the O1 strategy, only one element of the container is selected, and the selection is random. Is it possible to replace random selection with a more deterministic approach (e.g., selecting the first element)?

The problem with random selection is that it makes tests in CI less robust. For example, I enabled type-checking in some tests in my codebase, and they passed successfully. However, this was only because the random selection chose a "good" element from the container. The next day, my colleague changed an unrelated piece of code, altering the RNG state. As a result, a "bad" element was selected, causing Beartype to complain, and the test failed, leading to my colleague's frustration.

My codebase is large, with many people working on it. It would be great to be able to disable randomness via configuration to avoid situations like the one mentioned above :)</div>

**Comment by leycec** [comment]
**Superb request!** Yes. So much yes. You're right, of course. This is an excellent feature request. Deterministic CI is a lofty goal. @beartype should support you in that goal, @avolchek. You literally already resolved super-nasty issue #382. *Quid pro quo* and my genteel nature demands I do this for you and your justifiably frustrated colleague. When colleagues are frustrated with @beartype, @beartype has done something wrong. *Bad bear.* :bear: :angry: :bear: 

## This Plan: It Is a Good Plan

Let's add a new `BeartypeConf(strategy_O1_check_sequence_index: Optional[int] = None, ...)` option to @beartype's existing configuration API. Assuming you also use `beartype.claw` import hooks, you'd probably set this option like so:

```python
# In your "{your_package}.__init__" submodule:
from beartype import BeartypeConf
from beartype.claw import beartype_this_package
beartype_this_package(conf=BeartypeConf(strategy_O1_check_sequence_index=0))
```

The idea here is pretty simple. `strategy_O1_check_sequence_index`:

* Defaults to `None`, which then instructs @beartype to non-deterministically type-check one item of each sequence with a pseudo-random index. Understandably, you don't want that.
* Can be set to an integer, in which case @beartype will then deterministically type-check one item of each sequence with that index. You want that. Really, the only two sensible hard-coded indices are:
  * `strategy_O1_check_sequence_index=0`, type-checking the first item.
  * `strategy_O1_check_sequence_index=-1`, type-checking the last item.

This is (*...probably*) trivial stuff. So, this will (*...probably*) make it into our upcoming @beartype 0.19.0 release. Let's see if @leycec does anything at all! *So sleepy.* :yawning_face:

**Comment by leycec** [comment]
**WELCOME TO THE RESURRECTION OF #385.** Somehow, it has been two years. Nothing was done on this or most other issues, which have now exploded beyond the formerly unthinkable bounds of 100. You may now be thinking:

> "Say something I actually want to hear."

I was just *getting* to that part, silly. I'm immediately resurrecting this long-standing reasonable request I should have done a lifetime ago, but didn't, because video games and weed. Why? It's a blocker for the completely unrelated feature request #626: Leycec's Last Theorem of QA, The Issue So Brutal Its Official Mascot Is Now the Almost-Dead Doom Guy Finnadie. :finnadie: 

Wierd name, finnadie. I don't make up emoji names. I just blast them on GitHub. I shall now say something useful.

## I Have No Idea What I Wrote Above, But I'm Throwing It All Away

**Note to bros at home:** This is why you do not write paragraphs of throwaway monologue on GitHub. No one wants to read that. Not even the original author wants to read that.

Actually, I did read that. I lied. You know you're reading lies when you're reading a @leycec monologue. That's actually a great idea my younger self had. The option name `strategy_O1_check_sequence_index` is... uh, pretty rough. Even I hate that. We'll probably iterate there. Otherwise, that was a surprisingly sensible API.

Younger self, you rock! Thanks so much, bro. I have almost as much hair as you, but clearly less brain cells. Neural degradation is already setting in, huh? That's rough. Give a poor dev guy a break. 😓

**Comment by leycec** [comment]
On second thought... nobody needs or wants `strategy_O1_check_sequence_index`. Too heavyweight. Unpronounceable. Unreadable. Instead, we now rollin' with:

```python
BeartypeConf(is_random: bool = True)
```

Finally. Sanity. At last. 😮‍💨

**Comment by Glinte** [comment]
I can imagine someone in the future wanting a full DI of the rng function rather than just a True/False toggle, but there is no such person complaining yet so maybe pretend they don't exist

**Comment by leycec** [comment]
**Totally.** I too imagine a younger me with thin pursed lips that intone (icily, of course):

> "Your API is crude, your belly is fat, and... HEY. *Where did all your hair go, old guy*!?"

I sympathize with my jerkish younger self. But he might actually have been wrong about something.

A crude black box is probably preferable here. Why? Dumb efficiency. Always dumb. Always efficiency. As you'd expect because your are @Glinte, hard-coding `0` as the item to be type-checked generates dumb-fast code. The only other valid general-purpose item index would be `-1`, which... kinda seems pointless, I hope? Surely there's no valid reason for anyone to prefer the first over the last item? Gods. There probably is. Too bad, everyone! I'm ignoring your valid concerns in a desperate bid to release a working thing. 😆 

Thanks so for chiming in, @Glinte. You're always here with a helpful comment. You listen to a bald man grumble when no one else honestly cares. I really appreciate it. It's been a long winter. Like... it's *still* winter! How!? Honestly. These two layers of mohawk hair socks that are the only thing keeping my feet alive tell no lies.

**Comment by leycec** [comment]
**Aaaaaaaaand...** it's done (minus tests, which matter, but don't). Seriously!? Users have been asking for deterministic type-checking for a decade. Then it just takes me a single night to implement? Welp. Hopefully nobody ever reads this comment and realizes that putting an autist in charge may not have been the brightest idea. 🤣

**Comment by Glinte** [comment]
> hard-coding 0 as the item to be type-checked generates dumb-fast code. The only other valid general-purpose item index would be -1

That's why i'm not saying DI the index generator function, but DI the rng function. In very extreme cases (fuzzing?) it is possible that people want to use an rng that they can control and have reproducible errors.

> You listen to a bald man grumble when no one else honestly cares. I really appreciate it.

Thank you. I don't consider myself as someone who likes reading but I read like 3-4 hours every day, so the amount you write isn't really a lot for me. grumble is a funny word but nah

> It's been a long winter. Like... it's still winter!

Canada is a tough place huh

**Comment by leycec** [comment]
**Resolved by d28d4d0e9bc46275587ff.** What a farce this was to implement. Two years of doing nothing culminated in eight hours of reluctantly pounding on a keyboard. And even that wouldn't have happened if #626 hadn't made me. The things I do for @ilyapoz and his hat. 😂 

> In very extreme cases (fuzzing?) it is possible that people want to use an rng that they can control and have reproducible errors.

**Oh.** White-hat security. Super RNG-sensitive use cases where even the stock Mersene Twister fails to generate sufficient (or sufficiently constrained) entropy, huh? I tip the fedora as if I actually understood anything I just wrote.

Since we are now in the realm of white-hat security, we are in the realm of a trillion-dollar industry. Their budgets can spare a dollar. But they're not sparing even that for @beartype! @beartype thus insensitively chooses to fill out more US expat tax forms that serve no purpose whatsoever for anyone rather than doing something that helps everyone.

Curse those US expat tax forms. It's not like any US expats ever even pay taxes! The minimum foreign income thresholds are so high that the twelve living expats who actually *do* also live in homes with names like "Shambala II" and "Welcome to Xanadu Iceberg Palace." If we did, I'd be sitting on a cold Canadian bag of toonies, our bathtub would be filled with toonies, and the cats would wade through toonies on the way to the kibble bowl (also made of toonies).

Oh, Snow Gods. No one even knows what "toonies" means. My rant was ineffectual. Canadian culture has diverged so far from the human optimum we're basically all beavers at this point. 😭 

> I don't consider myself as someone who likes reading but I read like 3-4 hours every day,

You read more than a man who doesn't like reading should have to read. You are someone who life has tortured with reading. I see the deities have much to answer for.

**WAIT.** You don't like reading, you say...? Hah, hah. This is a great time to proselytize all my favourite scifi, fantasy, and Japanese lightnovel picks from the last eighty decades. Truly, the deities have blessed one of us!

Alright. I'll spare you. I lied. Can I interest you in a burgeoning genre called "LitRPG"? No?

Alright. I can accept rejection. But I'm still chuffling disconsolately to myself. 😑 <-- *@leycec chuffles*

> Canada is a tough place huh

![](https://media1.tenor.com/m/Amd82EBAABQAAAAC/toga-himiko-toga.gif)
<sup>*villainess toga himiko means well*</sup>

**Comment by JWCS** [comment]
The litrpg genre was originally novel and interesting, but I've come to realize any good work with such elements would be a better work without both author and reader staring at tables of numbers and pretending they mean something, or convey plot/stakes. I estimate the quality drops (for a decent author) 1/3-2/3 of a letter grade for litrpg elements, plus another 1/2-whole letter grade if isekai-ish. There's better ("classic") techniques for achieving the same thing, without the baggage the author didn't intend to drag in. (There are a _few_ works that don't suffer the drop, but they _fully_ embrace the implications, constraints, and baggage to go whole stock lock and barrel into it. If the author isn't that committed (or interested) to pulling it off, it's not worth.)

Thanks for the feature - this gets us half way to deterministic unit tests (the other half full fat ish checking), which for unit tests is the most pervasively missing feature of beartype. 
The conf name is also way nicer.

**Comment by leycec** [comment]
**You're so welcome!** `is_random` is, indeed, the only sane name. I stared down the barrel of 20 different alternatives. Every one reeked of "No one will ever use this, because the name sounds like it destroys the Casimir quantum force along with your codebase."

Full-fat `O(n)` type-checking is actually next up on @beartype's plate after `0.23.0`. It's time. It's years past time. Everyone's frustrated by that glaring omission. And I'm frustrated by everyone's frustration. It's transitive frustration. 😂 

> The litrpg genre was originally novel and interesting

**Yes! Yes!** A lit-fantasy debate by adulting devs emerges with grim finality on the @beartype issue tracker. Finally. I cherish this moment I have been waiting for. @posita! *GitHubbers are talking about fantasy here.*

My take is... distinctly autistic. Personally, I cannot *stand* modern Western fantasy. It's organizing principles are aligned with grimdark darkness, filth, and degeneracy. The villains all have plot armor. The heroes are all tortured and murdered within the first several pages. Everyone left is a hollow wreck of an anti-hero shambling through a maleficent political hellscape populated by cannibal pirate zombies. Genre subtags like "mindbreak," "mind control," and "corruption" are par for the course. Real-world examples include:

* **James Islington.** "The Shadow of What Was Lost." One of the worst fantasy novels I've ever tried to read. Billed as a Sanderson-like, I was immediately interested. It was nothing like Sanderson. The villains all have plot armor. The heroes are all tortured and murdered within the first several pages. Everyone left is a hollow wreck of an anti-hero shambling through a maleficent political hellscape populated by cannibal ~~pirate~~ mind-broken zombies. Seriously. When the cannibal mind-broken zombies showed up, I "noped!" out and regretted my life choices. Yet again. How many times must you connive and deceive me, Western publishing world!?
* **Michael J. Sullivan.** The Riyria Revelations. The only thing being revealed here is that modern fantasy authors could not even write their way out of a paper bag drenched in cat vomit. The main "heroes" are raving lunatics who repeatedly destroy their entire civilization by making the worst possible decisions, again... and again... *and again.* Riyria was as equally putrescent as it was predictable. More life decisions to regret. Great. Just what a bald middle-aged man needs.
* **Brent Weeks.** "Way of Shadows" and "Shadow's Edge" (Night Angel). *OMFFFFFFFFFFFG.* Brent Weeks insanely wrote himself into a corner filled with paper bags drenched in cat vomit by (*waitforit*) writing Superman into a Celtic fantasy trilogy. The main protag is actually even *more* powerful than Superman. You didn't think that was possible. Neither did I. Apparently, it is. The main protag is so powerful he literally cannot die. Ever. He just gets immediately resurrected. It's the Ultimate Dumb Healing Mechanic. And I know Dumb Healing Mechanics. I read JoJo's Bizarre Adventures. You'd think the main protag would wisely choose to punch bad guys given that he's literally a walking planet destroyer. Right? Nope. His girlfriend tells him not to punch bad guys. He stupidly listens. Then the bad guys destroy civilization. And then he *still* does nothing. It's the classic "This is fine" trope as a little dog pretends the house is not on fire. My Gods... I have never been angrier. That's a lie. I just did US expat taxes – the human activity so intrinsically angering it can only be done while listening to tinkly piano music.
* **Robin Hobb.** "Assassin's Apprentice" and "Royal Assassin" (Farseer Trilogy). Quite possibly the worst fantasy I have ever read. And you just know I've expurgated over bad modern fantasy. But Farseer... *wow.* It was breathtakingly bad. Allow me to reiterate the mantra I now take to bed with me every evening: "The villains all have plot armor. The heroes are all tortured and murdered within the first several pages. Everyone left is a hollow wreck of an anti-hero shambling through a maleficent political hellscape populated by cannibal pirate zombies." All of that stuff? Yeah. That's Farseer. The main protag (Fitz) is **profoundly mentally ill.** He's a nightmare fuel villain in a nightmare fuel world inhabited only by nightmare fuel villains. And cannibal pirate zombies. Gods. How could we forget *them*? In one especially rousing section of the first book, Fitz madly capers through the castle randomly stabbing passerby with a knife. Like... I don't even know. The dude was an unhinged bipolar trainwreck who openly attacks innocent humans. He miserably fails at everything he does, because randomly stabbing passerby with knives has a way of inculcating failure. He's the sort of happy-go-lucky plucky wolf whisperer who make modern cannibal pirate zombies look fashionable and upstanding by compare. He was *insufferable*. And I ask myself: "Why would anyone voluntarily subject themselves to aberrant mind-tripe that only increases the urge for self-harming masochism?" If I wanted to openly and publicly flagellate myself, I'd work in the modern office environment. What I would *not* do is ever touch Robin Hobb with a ten-foot tentacle ever again. Fool me once, shame on the publisher. Fool me twice, shame on the cannibal pirate zombies.

And I'm clearly not the only one. The entire for-profit industry of modern fantasy was on its vestigial sales deathbed, until **(A)** romantasy resurrected fantasy for female readers, **(B)** litRPG resurrected fantasy for male readers, and **(C)** Japanese lightnovels resurrected fantasy for those who'd prefer not to even specify a gender.

Miya Kazuki's "Ascendance of a Bookworm" is so fundamentally head-and-shoulders above every Western author in pretty much every genre that it's kinda depressing. Sanderson is the only Western author that's her equal – but even Sanderson's not quite up to par anymore. He spread himself thin. Quantity over quality is a bad look for anyone. He doesn't do comedy, although he tries. He's painfully unfunny, jaw-droppingly dead serious, and unquestionably epic in that Spinal Tap "This Amp Goes to 11!" kind of way. Sanderson lacks flexibility, in other words. He does one thing and one thing only. High epic fantasy. And when he steers away from his gore-splattered bread-and-butter into young adult fanfic gender swap territory? He pretty much sucks, guys. He can't hack anything except the very specific niche he occupies and dominates.

Miya Kazuki, on the other hand? This world has never seen her like. And may never again.

In other words, this autist has *opinions.* 😂
